### ISSUE TITLE
No member named socket_min_rcvbuf_size in defconfig.c error when building in buildroot

### ISSUE THREAD

**Issue by pscheie** [eclipse-cyclonedds/cyclonedds#1092]
When I try to build cyclonedds on my Linux Mint 19 machine, it builds just fine.  Now I'm trying to build it under buildroot, but I'm getting these two errors:

/home/petre/buildroot-2021.02.2.beaglebone/output/build/cyclonedds-eedb3b97272732135d773cc681ae5e496f5a682f/src/core/defconfig.c: In function ‘ddsi_config_init_default’:
/home/petre/buildroot-2021.02.2.beaglebone/output/build/cyclonedds-eedb3b97272732135d773cc681ae5e496f5a682f/src/core/defconfig.c:58:8: error: ‘struct ddsi_config’ has no member named ‘socket_min_rcvbuf_size’; did you mean ‘socket_rcvbuf_size’?
   58 |   cfg->socket_min_rcvbuf_size.isdefault = 1;
      |        ^~~~~~~~~~~~~~~~~~~~~~
      |        socket_rcvbuf_size
/home/petre/buildroot-2021.02.2.beaglebone/output/build/cyclonedds-eedb3b97272732135d773cc681ae5e496f5a682f/src/core/defconfig.c:59:8: error: ‘struct ddsi_config’ has no member named ‘socket_min_sndbuf_size’; did you mean ‘socket_sndbuf_size’?
   59 |   cfg->socket_min_sndbuf_size = UINT32_C (65536);
      |        ^~~~~~~~~~~~~~~~~~~~~~
      |        socket_sndbuf_size

Looking at ddsi_config.h, I see that those two members are indeed not in there.  But if that's the case, why does it build for me on my host system? 

I was previously able to build 0.8.1 under buildroot (downloading the tar file), but I just switched to having BR use git to pull the latest commits (which is "version 0.9.0", right?).  But when I made that switch, I started getting the above errors.

**Comment by eboasson** [comment]
If I were to guess ... `defconfig.c` is generated during the build and (for whatever reason) CMake didn't find it necessary to regenerate it when you switched branches. (The old `socket_min_rcvbuf_size` morphed into `socket_rcvbuf_size` (likewise for send buffers). See commit 8da20663feef072e0d6f0092467f30fb62d8124b.)

**Comment by pscheie** [comment]
Hmm.  I created an entirely new buildroot environment (v2021-02.8), updated the package make file for CMake (i.e., cmake.mk) from 3.15.5 to 3.18.6, and that all successfully builds.  Then I added the package for cyclonedds which uses version 0.9.0a1 pulled from https://github.com/eclipse-cyclonedds/cyclonedds/archive/refs/tags/0.9.0a1.tar.gz.  But when I run make, I'm still getting errors that defconfig.c is specifying socket_min_rcvbuf_size and socket_min_sndbuf_size, even though in this case there was no previous defconfig.c that those could have carried over from.

**Comment by eboasson** [comment]
Well, that means a new hypothesis is needed ...

Here is one: it looks like the `ddsconf` build-tool that generates (among other things) the `defconfig.c` file gets installed. It shouldn't be, it serves no purpose outside the Cyclone DDS build, I'll do a PR to stop it from being installed in the future.

It does mean that if you have 0.8 installed somewhere, it could include `ddsconf`, and if that is in the path, perhaps it is the old one that would get executed instead of the freshly built one. Should be easy to find out ... `which ddsconf`, and if it exists (and the one found is indeed Cyclone DDS's one) delete it and the generated `defconfig.c` and try again.

Hopefully is this simple🤞

**Comment by pscheie** [comment]
Well nuts, I thought you had nailed it: I *did* have a ddsconf in my $PATH.  The directory, /opt/cyclonedds/bin, originally had 0.8.1 installed, although I think I installed 0.9.0a1 over it (although that doesn't always overwrite everything, might still be some relics lying around).  Regardless, I took /opt/cyclonedds out of my $PATH (and verified that there was no other ddsconf in the path) and tried building again (make clean all).   And...defconfig.c is still using socket_min_rcvbuf_size. :-(

I don't know much about CMake, but how is defconfig.c constructed?  That is, is there some sort of base template it starts with that might be out of date?

**Comment by pscheie** [comment]
Just an update that may or may not be significant:  As mentioned in my previous post, /opt/cyclonedds/bin, where ddsconf lived, was in my path, so I redefined the path without that, but still got the same defconfig.c problem.  But then I wondered if somehow say, /opt/cyclonedds/lib might be messing with things.  So, I renamed /opt/cyclonedds to /opt/cyclonedds.x, so nothing would have any sort of pre-existing path to the host's cyclonedds.  Ran 'make clean; make all', and now it fails because it cannot find ddsconf at all:

CMake Error at src/tools/ddsconf/CMakeLists.txt:60 (find_program):
  Could not find _executable using the following names: ddsconf

I recall thinking earlier (and had forgotten until this moment) that it was going to be necessary to install cyclonedds on the build host in order to be able to cross-compile under buildroot, right? 

I'll run 'ddsconf -f defconfig' and report back.

**Comment by pscheie** [comment]
The output from `ddsconf -f defconfig`, using the ddsconf in my /opt/cyclonedds/bin directory, had the old socket_min_rcvbuf_size elements; so something is wrong/old in /opt/cyclonedds.  Fine, get rid of that, re-downloaded https://github.com/eclipse-cyclonedds/cyclonedds/archive/refs/tags/0.9.0a1.tar.gz (so as to be as similar to what BR does), compiled & installed, and then the output from `ddsconf -f defconfig' showed the updated elements, e.g., socket_rcvbuf_size.min, etc.

So, with this new ddsconf in my $PATH, I went back into buildroot env and ran `make clean all`.  It's still failing, but it got past the socket_rcvbuf* items, and now it's failing because of missing ssl elements:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c: In function ddsi_config_init_default’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:86:6: error: ‘struct ddsi_config’ has no member named ‘ssl_verify’
   86 |   cfg->ssl_verify = INT32_C (1);
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:87:6: error: ‘struct ddsi_config’ has no member named ‘ssl_verify_client’
   87 |   cfg->ssl_verify_client = INT32_C (1);
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:88:6: error: ‘struct ddsi_config’ has no member named ‘ssl_keystore’
   88 |   cfg->ssl_keystore = "keystore";
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:89:6: error: ‘struct ddsi_config’ has no member named ‘ssl_key_pass’
   89 |   cfg->ssl_key_pass = "secret";
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:90:6: error: ‘struct ddsi_config’ has no member named ‘ssl_ciphers’
   90 |   cfg->ssl_ciphers = "ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH";
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:91:6: error: ‘struct ddsi_config’ has no member named ‘ssl_rand_file’
   91 |   cfg->ssl_rand_file = "";
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:92:6: error: ‘struct ddsi_config’ has no member named ‘ssl_min_version’
   92 |   cfg->ssl_min_version.major = 1;
      |      ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/cyclonedds-0.9.0a1/src/core/defconfig.c:93:6: error: ‘struct ddsi_config’ has no member named ‘ssl_min_version’
   93 |   cfg->ssl_min_version.minor = 3;

I didn't set `ENABLE_SSL=NO` (since I will need SSL eventually); not sure if that's related or not.

**Comment by eboasson** [comment]
If this is a cross-build by any chance? Because if the host has SSL and the target doesn't have SSL, and my suspicion that the host's configuration is used for the target's `defconfig.c` is correct, then you would expect this on the target.

That would be a bug in the cross-build setup. I haven't personally done any cross-builds with Cyclone (or CMake, more generally) but I'm sure that's a problem we can find a solution for. The temporary workaround would then be to set `ENABLE_SSL=NO`.

**Comment by thijsmie** [comment]
I've skipped over this thread before but just re-read it and indeed what @eboasson suggests should work. The crossbuilds are still somewhat painful since a _native_ ddsconf must generate a _target_ defconfig, which only works if all the relevant CMAKE options are matching. This really only can be solved by turning the ddsconf build into a nested CMAKE invocation instead of using one already available, to avoid this kind of mysterious mismatching errors. It is definitely on my TODO to tackle this, simply haven't had the time.

**Comment by pscheie** [comment]
Yes, this is a cross compile.  Host system is x86_64, target is ARM (BeagleBoneBlack).  I appended the argument -DENABLE_SSL=NO (based on my skimming of the CMake man page, I think that's the right syntax) to the `-DCMAKE_INSTALL_PREFIX=$(CYCLONEDDS_TARGET_DIR) -DBUILD_EXAMPLES=ON' value for the CYCLONEDDS_CONF_OPTS parameter in the cyclonedds.mk file for the cyclonedds package in buildroot; but when I run 'make clean all', it still fails over the "missing" ssl_* members in struct ddsi_config.  So, it appears I'm still doing something wrong.

**Comment by pscheie** [comment]
> Well, that means a new hypothesis is needed ...
> 
> Here is one: it looks like the `ddsconf` build-tool that generates (among other things) the `defconfig.c` file gets installed. It shouldn't be, it serves no purpose outside the Cyclone DDS build, I'll do a PR to stop it from being installed in the future.
> 
So, installing cyclonedds under buildroot should not require cyclonedds to be installed on the host system beforehand, right?  If so, something is wrong, because when I remove the host's cyclonedds (which is outside the buildroot directory), buildroot gets an error that it can't find ddsconf when building cyclonedds.  IOW, it appears cyclonedds is *not* installing or providing its own ddsconf when building under buildroot.

As mentioned, this is a cross-compile.  Does that mean ddsconf has to be compiled for the target architecture (Arm)?

**Comment by pscheie** [comment]
Update: I am now able to get cyclonedds 0.9.0a1 to build under buildroot cross-compiled for Arm.  I had to add OpenSSL as a dependency in the .mk file for cyclonedds (which I wanted anyway) because the external ddsconf, which building uses, was compiled with OpenSSL support.  So, setting ENABLE_SSL=NO does not seem to work if ddsconf was built with SSL support.  

But do I understand correctly that I shouldn't need to rely on an external ddsconf?  That is, it should build its own? Or is this a cross-compile problem?

Anyway, while it does now build, the install fails because docs/cmake_install.cmake at line 46 is looking for a ../LICENSE file, but there isn't one (there's a license.txt).  It also looks for some others, successfully, such as README.md, fwiw.  This seems like a typo/bug in cmake_install.cmake, although I don't get this problem when installing natively outside of buildroot. 

Suggestions?

**Comment by pscheie** [comment]
One more update: I copied license.txt to LICENSE and then the install succeeded.  So, I know have a workable cyclonedds in buildroot.  

This issue, #1092, should arguably be closed because the problem as originally posted, socket_min_rcvbuf_size in defconfig.c, was resolved by properly updating my host installation of cyclonedds.  However, the license.txt/LICENSE issue mentioned just above still needs attention, i.e., the file should be renamed or CMakeLists.txt should be changed to look for the correct name.  I'm still not clear as to why installing cyclonedds in buildroot seems to require having cyclonedds first installed on the host system since building cyclonedds generates its own ddsconf as I understand it (or perhaps I don't).
